controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java

```diff
@@ -100,23 +101,25 @@ public ControllerEventProcessors(final String host,
                                      final ConnectionFactory connectionFactory,
                                      final StreamMetadataTasks streamMetadataTasks,
                                      final StreamTransactionMetadataTasks streamTransactionMetadataTasks,
-                                     final ScheduledExecutorService executor) {
+                                     final ScheduledExecutorService executor,
+                                     final TransactionMetrics transactionMetrics) {
```

Comment:
Should we pass new transactionMetrics object here? Can we not do it like other metrics, where we use static provider and get specific transaction metrics?

Reply:
I see two options:
1) We pass the metrics object as currently done, so we need to modify the constructor of all classes using it.
2) We make `TransactionMetrics` (and maybe `StreamMetrics` too) singleton objects, so we only need to invoke something like `TransactionMetrics.getInstance().xxxx()` from the places where we want to report metrics. This would not require to change the constructors for that class.

Please, let me know your opinion on that.

Reply:
My preference would be to use state/singleton objects. So let us focus on `2` and see if it is a viable solution. 
Having static/singleton objects exposed via static getters, they can be accessed from anywhere without having to pass these objects in constructors. 
However, do remember that controller modules can be reinitialized (in case of zk session expiration, all modules restart internally).
We need to assess if singleton metrics objects could have any detrimental effect in any way. Let me elaborate - if there was any state (any metrics) in the previous iteration of metrics object, do we need to reset such states during restarts of other modules that use this metrics object?

If not, we can use singleton objects without any worry.  
However, if we need to instantiate new metrics object each time we restart, then we could have these metrics initialized in ControllerServiceStarter and reinitialized upon each restart which refreshes their state or creates a new object altogether perhaps.